### PR TITLE
DT-1232: 🚧 Start to add in service based mfa

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/UserSelectorAuthorizationEndpoint.kt [hunk 1]
```diff
@@ -32,13 +32,14 @@ class UserSelectorAuthorizationEndpoint(
   private val userRetriesService: UserRetriesService,
   private val telemetryClient: TelemetryClient,
 ) {
-  @GetMapping(value = ["/oauth/authorize"])
+  @GetMapping("/oauth/authorize")
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/UserSelectorAuthorizationEndpoint.kt:35]: just a bit of tidying up

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt [hunk 2]
```diff
@@ -32,25 +33,20 @@ class MfaService(
   private val userRetriesService: UserRetriesService,
 ) {
 
-  private val ipMatchers: List<IpAddressMatcher>
-
-  init {
-    ipMatchers = whitelist.map { ip -> IpAddressMatcher(ip) }
-  }
+  private val ipMatchers: List<IpAddressMatcher> = whitelist.map { ip -> IpAddressMatcher(ip) }
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt:36]: will change this to allowlist in next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt [hunk 3]
```diff
@@ -32,25 +33,20 @@ class MfaService(
...
 
   companion object {
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  fun needsMfa(authorities: Collection<GrantedAuthority>): Boolean {
-    // if they're whitelisted then no mfa
+  fun outsideApprovedNetwork(): Boolean {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt:42]: made separate function so can call independently

**Review by JamieDraperUK — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt [hunk 4]
```diff
@@ -32,25 +33,20 @@ class MfaService(
...
     val ip = IpAddressHelper.retrieveIpFromRequest()
-    return if (ipMatchers.any { it.matches(ip) }) {
-      false
-      // otherwise check that they have a role that requires mfa
-    } else authorities.stream().map { it.authority }.anyMatch { r -> mfaRoles.contains(r) }
+    return !ipMatchers.any { it.matches(ip) }
   }
 
+  fun needsMfa(authorities: Collection<GrantedAuthority>): Boolean =
+    outsideApprovedNetwork() && authorities.stream().map { it.authority }.anyMatch { r -> mfaRoles.contains(r) }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt:48]: I don't think you need a stream here.

**Review by mikehalmamoj — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt (hunk 4 above)
Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt:48]: agreed, left over from java land - will fix in subsequent PR

**Pull request merged**